app/src/main/java/com/github/doomsdayrs/apps/shosetsu/backend/TarArchive.kt

```diff
@@ -0,0 +1,66 @@
+package com.github.doomsdayrs.apps.shosetsu.backend
+
+import org.kamranzafar.jtar.*
+import java.io.File
+import java.io.InputStream
+import java.io.RandomAccessFile
+
+class TarArchive(file:String):File(file) {
+
+
+    fun read(name: String): InputStream {
+        val tar = TarInputStream(this.inputStream())
+
+        generateSequence { tar.nextEntry }.find {  it.name.contains(name,true) }
+
+        val buffer = tar.readBytes()
+        tar.close()
+        return buffer.inputStream()
+    }
+
+
+
+    fun write(name: String, buffer: ByteArray) {
+        val t = TarOutputStream(this, true)
+        val header = TarHeader()
+
+        header.name = StringBuffer(name)
+        header.size = buffer.size.toLong()
+        t.putNextEntry(TarEntry(header))
+        t.write(buffer)
+        t.close()
+    }
+
+
+
+    fun delete(filename: String) {
+        val tar = TarInputStream(this.inputStream())
+        val offset: Long
+        val nextOffset:Long
+        val entry =  generateSequence { tar.nextEntry }.find { it.name.contains(filename,true) }
+
+        if (entry!=null){
+            offset = tar.currentOffset - TarConstants.HEADER_BLOCK
+            tar.nextEntry
+            nextOffset = tar.currentOffset - TarConstants.HEADER_BLOCK
+
+            tar.close()
+
+            val buffer = ByteArray((this.length() - nextOffset).toInt())
+            val raf = RandomAccessFile(this, "rw")
+
+            raf.seek(nextOffset)
+            raf.read(buffer)
+            raf.seek(offset)
+            raf.write(buffer)
+            raf.setLength(this.length() - (nextOffset - offset))
+            raf.close()
+        }
+        tar.close()
```

Comment:
Possibly closing an already closed stream. remove the previous call to close.

Reply:
But then  it will open Random access file And inputstream together.